### PR TITLE
s4878126 - VQVAE Trained over the HipMRI Study for Prostate Cancer Radiotherapy

### PR DESCRIPTION
**Pull request by aditya48781262**
_Aditya Sant - s4878126_

A _**VQVAE**_ is an expanded version of the variational autoencoder architecture for generative models, adding an extra layer for storing embeddings of discrete representations of encoded data. This model is No.7 on the list of available projects, falling under a hard difficulty level. 

After training on a consumer or a data center GPU for 50 epochs and running inference, the model achieves an SSIM score of 0.68, marginally crossing the specified SSIM threshold on the task sheet. 

This version was implemented in PyTorch using the object-orientated programming (OOP) paradigm, with different layers (encoder,quantisation,decoder) and the model itself represented through objects and classes. 

ADAM was used as the final optimiser, with a learning rate that be adjusted between 2e-4 and 2e-5 and a batch size of 64. A full description of the model, training, theory and results can be found in the README.md file.

Directory structure:
/ **topic-recognition** **(head directory)**
// **HipMRI-VQVAE-s4878126**
/// **results** (contains images and generated scans)
/// **scripts** (contains slurm script for training and running inference on Rangpur)
/// **modules.py** (contains encoder, quantisation and decoder architecture, compiles into a VQVAE within train.py)
/// **dataset.py** (fetches training, validation and testing data from UQ Rangpur, or path can be changed to local if dataset is stored locally)
/// **utils.py** (contains function provided for fetching 2D data in NIFTI format; function is slightly modified, please refer to README.md. Additionally, parameters can also be modified in this file)
/// **train.py** (contains a training loop for 50 epochs)
/// **predict.py** (runs inference for generating images from test dataset)

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Reasonably clear and resembling reconstructed images, with an SSIM over 0.6 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

4. Design:
Modularized design and good practice using utils script (1 Mark)

5. Commenting:
Sufficient docstrings and comments (1 Mark)

6. Algorithm above Normal Difficulty: (5 Marks)

7. Algorithm is Hard difficulty: (5 Marks)

**Suggestions**:

- It's really nice to have progressive experiments and hyperparameter tuning recorded with consequent losses :). You could also record SSIM in the same way;
- Training and predicting could be more modularized, but this is fine;
- No need to delete recognition/README.md;
- No validation used, could add into `train.py`.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Validation would be nice. -1

**Comment by shakes76** [comment]
Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.

**Comment by aditya48781262** [comment]
> Cant merge because of conflicting changes to main repo files. Please update for merge, doesn't affect grade.

@shakes76 I've updated the readme and license to match the topic-recognition branch's files. can you please try again
